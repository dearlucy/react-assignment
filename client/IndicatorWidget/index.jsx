import React, { Component } from "react";
import moment from "moment";
import style from "./style.less";
import { BarChart, Bar, CartesianGrid, XAxis, YAxis, Tooltip } from "recharts";

class Indicator extends Component {
  constructor() {
    super();
    this.state = {
      dataSet: [],
    };
  }
  componentDidMount() {
    const currMonth = moment().format("YYYY-MM");
    let url = `http://localhost:8000/series?from=${currMonth}&to=${currMonth}`;
    fetch(url)
      .then((response) => response.json())
      .then((data) => {
        this.setState({
          dataSet: data,
        })
      })
      .catch((error) => {
        console.error('Error:', error);
      });
  }

  render() {
    return (
      <div className={style.container}>
        <h3 className={style.title}>Values for the current month </h3>
        <BarChart
          width={300}
          height={300}
          data={this.state.dataSet}
          margin={{
            top: 5,
            right: 30,
            left: 20,
            bottom: 5,
          }}
        >
          <CartesianGrid strokeDasharray="2 2" />
          <XAxis dataKey="date" />
          <YAxis />
          <Tooltip />
          <Bar dataKey="value" fill="#8884d8" />
        </BarChart>
      </div>
    );
  }
}

export default Indicator;
