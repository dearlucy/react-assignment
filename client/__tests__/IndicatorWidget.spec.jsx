import React from "react";
import { shallow } from "enzyme";
import IndicatorWidget from "../IndicatorWidget";
import { BarChart, Bar, CartesianGrid, XAxis, YAxis, Tooltip } from "recharts";

describe("IndicatorWidget component", () => {
  let wrapper = shallow(<IndicatorWidget />);
  it("Should render component", () => {
    expect(wrapper).toMatchSnapshot();
  });
});

describe("IndicatorWidget renders components", () => {
  let wrapper;

  beforeEach(() => (wrapper = shallow(<IndicatorWidget />)));

  it("renders title", () => {
    expect(wrapper.find("h3").length).toBe(1);
  });

  it("Renders BarChart", () => {
    expect(wrapper.find(BarChart).length).toBe(1);
  });

  it("Renders Bar", () => {
    expect(wrapper.find(Bar).length).toBe(1);
  });

  it("Renders CartesianGrid", () => {
    expect(wrapper.find(CartesianGrid).length).toBe(1);
  });

  it("Renders XAxis", () => {
    expect(wrapper.find(XAxis).length).toBe(1);
  });

  it("Renders YAxis", () => {
    expect(wrapper.find(YAxis).length).toBe(1);
  });

  it("Renders Tooltip", () => {
    expect(wrapper.find(Tooltip).length).toBe(1);
  });

});


const componentDidMountSpy = jest.spyOn(IndicatorWidget.prototype, 'componentDidMount');

describe("Lifecycle method", () => {
  let wrapper = shallow(<IndicatorWidget />, { lifecycleExperimental: true });
  it("componentDidMoun called only once", () => {
      expect(componentDidMountSpy).toHaveBeenCalledTimes(1);
  })

  it("componentDidMoun not called twice", () => {
    expect(componentDidMountSpy).not.toHaveBeenCalledTimes(2);
  })
})