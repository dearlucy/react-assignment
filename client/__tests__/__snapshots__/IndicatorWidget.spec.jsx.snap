// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IndicatorWidget component Should render component 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className={undefined}
>
    <h3
        className={undefined}
    >
        Values for the current month 
    </h3>
    <BarChart
        barCategoryGap="10%"
        barGap={4}
        data={Array []}
        height={300}
        layout="horizontal"
        margin={
            Object {
                "bottom": 5,
                "left": 20,
                "right": 30,
                "top": 5,
              }
        }
        reverseStackOrder={false}
        stackOffset="none"
        width={300}
    >
        <CartesianGrid
            fill="none"
            horizontal={true}
            horizontalFill={Array []}
            horizontalPoints={Array []}
            stroke="#ccc"
            strokeDasharray="2 2"
            vertical={true}
            verticalFill={Array []}
            verticalPoints={Array []}
        />
        <XAxis
            allowDataOverflow={false}
            allowDecimals={true}
            allowDuplicatedCategory={true}
            dataKey="date"
            domain={
                Array [
                    0,
                    "auto",
                  ]
            }
            height={30}
            hide={false}
            mirror={false}
            orientation="bottom"
            padding={
                Object {
                    "left": 0,
                    "right": 0,
                  }
            }
            reversed={false}
            scale="auto"
            tickCount={5}
            type="category"
            width={0}
            xAxisId={0}
        />
        <YAxis
            allowDataOverflow={false}
            allowDecimals={true}
            allowDuplicatedCategory={true}
            domain={
                Array [
                    0,
                    "auto",
                  ]
            }
            height={0}
            hide={false}
            mirror={false}
            orientation="left"
            padding={
                Object {
                    "bottom": 0,
                    "top": 0,
                  }
            }
            reversed={false}
            scale="auto"
            tickCount={5}
            type="number"
            width={60}
            yAxisId={0}
        />
        <Tooltip
            active={false}
            allowEscapeViewBox={
                Object {
                    "x": false,
                    "y": false,
                  }
            }
            animationDuration={400}
            animationEasing="ease"
            contentStyle={Object {}}
            coordinate={
                Object {
                    "x": 0,
                    "y": 0,
                  }
            }
            cursor={true}
            cursorStyle={Object {}}
            filterNull={true}
            isAnimationActive={true}
            itemStyle={Object {}}
            labelStyle={Object {}}
            offset={10}
            separator=" : "
            useTranslate3d={false}
            viewBox={
                Object {
                    "x1": 0,
                    "x2": 0,
                    "y1": 0,
                    "y2": 0,
                  }
            }
            wrapperStyle={Object {}}
        />
        <Bar
            animationBegin={0}
            animationDuration={400}
            animationEasing="ease"
            data={Array []}
            dataKey="value"
            fill="#8884d8"
            hide={false}
            isAnimationActive={true}
            layout="vertical"
            legendType="rect"
            minPointSize={0}
            onAnimationEnd={[Function]}
            onAnimationStart={[Function]}
            xAxisId={0}
            yAxisId={0}
        />
    </BarChart>
</div>,
  "nodes": Array [
    <div
      className={undefined}
>
      <h3
            className={undefined}
      >
            Values for the current month 
      </h3>
      <BarChart
            barCategoryGap="10%"
            barGap={4}
            data={Array []}
            height={300}
            layout="horizontal"
            margin={
                  Object {
                        "bottom": 5,
                        "left": 20,
                        "right": 30,
                        "top": 5,
                      }
            }
            reverseStackOrder={false}
            stackOffset="none"
            width={300}
      >
            <CartesianGrid
                  fill="none"
                  horizontal={true}
                  horizontalFill={Array []}
                  horizontalPoints={Array []}
                  stroke="#ccc"
                  strokeDasharray="2 2"
                  vertical={true}
                  verticalFill={Array []}
                  verticalPoints={Array []}
            />
            <XAxis
                  allowDataOverflow={false}
                  allowDecimals={true}
                  allowDuplicatedCategory={true}
                  dataKey="date"
                  domain={
                        Array [
                              0,
                              "auto",
                            ]
                  }
                  height={30}
                  hide={false}
                  mirror={false}
                  orientation="bottom"
                  padding={
                        Object {
                              "left": 0,
                              "right": 0,
                            }
                  }
                  reversed={false}
                  scale="auto"
                  tickCount={5}
                  type="category"
                  width={0}
                  xAxisId={0}
            />
            <YAxis
                  allowDataOverflow={false}
                  allowDecimals={true}
                  allowDuplicatedCategory={true}
                  domain={
                        Array [
                              0,
                              "auto",
                            ]
                  }
                  height={0}
                  hide={false}
                  mirror={false}
                  orientation="left"
                  padding={
                        Object {
                              "bottom": 0,
                              "top": 0,
                            }
                  }
                  reversed={false}
                  scale="auto"
                  tickCount={5}
                  type="number"
                  width={60}
                  yAxisId={0}
            />
            <Tooltip
                  active={false}
                  allowEscapeViewBox={
                        Object {
                              "x": false,
                              "y": false,
                            }
                  }
                  animationDuration={400}
                  animationEasing="ease"
                  contentStyle={Object {}}
                  coordinate={
                        Object {
                              "x": 0,
                              "y": 0,
                            }
                  }
                  cursor={true}
                  cursorStyle={Object {}}
                  filterNull={true}
                  isAnimationActive={true}
                  itemStyle={Object {}}
                  labelStyle={Object {}}
                  offset={10}
                  separator=" : "
                  useTranslate3d={false}
                  viewBox={
                        Object {
                              "x1": 0,
                              "x2": 0,
                              "y1": 0,
                              "y2": 0,
                            }
                  }
                  wrapperStyle={Object {}}
            />
            <Bar
                  animationBegin={0}
                  animationDuration={400}
                  animationEasing="ease"
                  data={Array []}
                  dataKey="value"
                  fill="#8884d8"
                  hide={false}
                  isAnimationActive={true}
                  layout="vertical"
                  legendType="rect"
                  minPointSize={0}
                  onAnimationEnd={[Function]}
                  onAnimationStart={[Function]}
                  xAxisId={0}
                  yAxisId={0}
            />
      </BarChart>
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <Indicator />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": Indicator {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {},
        "refs": Object {},
        "state": Object {
          "dataSet": Array [],
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className={undefined}
>
          <h3
                    className={undefined}
          >
                    Values for the current month 
          </h3>
          <BarChart
                    barCategoryGap="10%"
                    barGap={4}
                    data={Array []}
                    height={300}
                    layout="horizontal"
                    margin={
                              Object {
                                        "bottom": 5,
                                        "left": 20,
                                        "right": 30,
                                        "top": 5,
                                      }
                    }
                    reverseStackOrder={false}
                    stackOffset="none"
                    width={300}
          >
                    <CartesianGrid
                              fill="none"
                              horizontal={true}
                              horizontalFill={Array []}
                              horizontalPoints={Array []}
                              stroke="#ccc"
                              strokeDasharray="2 2"
                              vertical={true}
                              verticalFill={Array []}
                              verticalPoints={Array []}
                    />
                    <XAxis
                              allowDataOverflow={false}
                              allowDecimals={true}
                              allowDuplicatedCategory={true}
                              dataKey="date"
                              domain={
                                        Array [
                                                  0,
                                                  "auto",
                                                ]
                              }
                              height={30}
                              hide={false}
                              mirror={false}
                              orientation="bottom"
                              padding={
                                        Object {
                                                  "left": 0,
                                                  "right": 0,
                                                }
                              }
                              reversed={false}
                              scale="auto"
                              tickCount={5}
                              type="category"
                              width={0}
                              xAxisId={0}
                    />
                    <YAxis
                              allowDataOverflow={false}
                              allowDecimals={true}
                              allowDuplicatedCategory={true}
                              domain={
                                        Array [
                                                  0,
                                                  "auto",
                                                ]
                              }
                              height={0}
                              hide={false}
                              mirror={false}
                              orientation="left"
                              padding={
                                        Object {
                                                  "bottom": 0,
                                                  "top": 0,
                                                }
                              }
                              reversed={false}
                              scale="auto"
                              tickCount={5}
                              type="number"
                              width={60}
                              yAxisId={0}
                    />
                    <Tooltip
                              active={false}
                              allowEscapeViewBox={
                                        Object {
                                                  "x": false,
                                                  "y": false,
                                                }
                              }
                              animationDuration={400}
                              animationEasing="ease"
                              contentStyle={Object {}}
                              coordinate={
                                        Object {
                                                  "x": 0,
                                                  "y": 0,
                                                }
                              }
                              cursor={true}
                              cursorStyle={Object {}}
                              filterNull={true}
                              isAnimationActive={true}
                              itemStyle={Object {}}
                              labelStyle={Object {}}
                              offset={10}
                              separator=" : "
                              useTranslate3d={false}
                              viewBox={
                                        Object {
                                                  "x1": 0,
                                                  "x2": 0,
                                                  "y1": 0,
                                                  "y2": 0,
                                                }
                              }
                              wrapperStyle={Object {}}
                    />
                    <Bar
                              animationBegin={0}
                              animationDuration={400}
                              animationEasing="ease"
                              data={Array []}
                              dataKey="value"
                              fill="#8884d8"
                              hide={false}
                              isAnimationActive={true}
                              layout="vertical"
                              legendType="rect"
                              minPointSize={0}
                              onAnimationEnd={[Function]}
                              onAnimationStart={[Function]}
                              xAxisId={0}
                              yAxisId={0}
                    />
          </BarChart>
</div>,
        "_debugID": 2,
        "_renderedOutput": <div
          className={undefined}
>
          <h3
                    className={undefined}
          >
                    Values for the current month 
          </h3>
          <BarChart
                    barCategoryGap="10%"
                    barGap={4}
                    data={Array []}
                    height={300}
                    layout="horizontal"
                    margin={
                              Object {
                                        "bottom": 5,
                                        "left": 20,
                                        "right": 30,
                                        "top": 5,
                                      }
                    }
                    reverseStackOrder={false}
                    stackOffset="none"
                    width={300}
          >
                    <CartesianGrid
                              fill="none"
                              horizontal={true}
                              horizontalFill={Array []}
                              horizontalPoints={Array []}
                              stroke="#ccc"
                              strokeDasharray="2 2"
                              vertical={true}
                              verticalFill={Array []}
                              verticalPoints={Array []}
                    />
                    <XAxis
                              allowDataOverflow={false}
                              allowDecimals={true}
                              allowDuplicatedCategory={true}
                              dataKey="date"
                              domain={
                                        Array [
                                                  0,
                                                  "auto",
                                                ]
                              }
                              height={30}
                              hide={false}
                              mirror={false}
                              orientation="bottom"
                              padding={
                                        Object {
                                                  "left": 0,
                                                  "right": 0,
                                                }
                              }
                              reversed={false}
                              scale="auto"
                              tickCount={5}
                              type="category"
                              width={0}
                              xAxisId={0}
                    />
                    <YAxis
                              allowDataOverflow={false}
                              allowDecimals={true}
                              allowDuplicatedCategory={true}
                              domain={
                                        Array [
                                                  0,
                                                  "auto",
                                                ]
                              }
                              height={0}
                              hide={false}
                              mirror={false}
                              orientation="left"
                              padding={
                                        Object {
                                                  "bottom": 0,
                                                  "top": 0,
                                                }
                              }
                              reversed={false}
                              scale="auto"
                              tickCount={5}
                              type="number"
                              width={60}
                              yAxisId={0}
                    />
                    <Tooltip
                              active={false}
                              allowEscapeViewBox={
                                        Object {
                                                  "x": false,
                                                  "y": false,
                                                }
                              }
                              animationDuration={400}
                              animationEasing="ease"
                              contentStyle={Object {}}
                              coordinate={
                                        Object {
                                                  "x": 0,
                                                  "y": 0,
                                                }
                              }
                              cursor={true}
                              cursorStyle={Object {}}
                              filterNull={true}
                              isAnimationActive={true}
                              itemStyle={Object {}}
                              labelStyle={Object {}}
                              offset={10}
                              separator=" : "
                              useTranslate3d={false}
                              viewBox={
                                        Object {
                                                  "x1": 0,
                                                  "x2": 0,
                                                  "y1": 0,
                                                  "y2": 0,
                                                }
                              }
                              wrapperStyle={Object {}}
                    />
                    <Bar
                              animationBegin={0}
                              animationDuration={400}
                              animationEasing="ease"
                              data={Array []}
                              dataKey="value"
                              fill="#8884d8"
                              hide={false}
                              isAnimationActive={true}
                              layout="vertical"
                              legendType="rect"
                              minPointSize={0}
                              onAnimationEnd={[Function]}
                              onAnimationStart={[Function]}
                              xAxisId={0}
                              yAxisId={0}
                    />
          </BarChart>
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Indicator />,
}
`;
